frontend/src/reducer/workspaceList.js

```diff
@@ -33,19 +36,28 @@ export function workspaceList (state = [], action) {
         memberList: []
       }))
 
+    case `${ADD}/${WORKSPACE_LIST}`:
+      return [
+        ...state,
+        ...action.workspaceList
+          .filter(w => !state.some(s => s.id === w.workspace_id))
+          .map(ws => ({
+            ...serialize(ws, serializeWorkspaceListProps),
+            sidebarEntryList: ws.sidebar_entries.map(sbe => serialize(sbe, serializeSidebarEntryProps)),
+            memberList: []
+          }))
+      ].sort(sortWorkspaceList)
+
+    case `${REMOVE}/${WORKSPACE_LIST}`:
+      return state.filter(ws => ws.id !== action.workspace.workspace_id)
+
     case `${SET}/${WORKSPACE_LIST}/isOpenInSidebar`:
       return state.map(ws => ({ ...ws, isOpenInSidebar: ws.id === action.workspaceId ? action.isOpenInSidebar : ws.isOpenInSidebar }))
 
     case `${SET}/${WORKSPACE_LIST_MEMBER}`:
       return state.map(ws => ({
         ...ws,
-        memberList: action.workspaceListMemberList.find(wlml => wlml.workspaceId === ws.id).memberList.map(m => ({
-          id: m.user_id,
-          publicName: m.user.public_name,
-          role: m.role,
-          isActive: m.is_active,
-          doNotify: m.do_notify
-        }))
+        memberList: action.workspaceListMemberList.find(wlml => wlml.workspaceId === ws.id).memberList.map(m => (serializeMember(m)))
```

Comment:
you should not use this pre refactor function. It'll be deprecated.
Use the new `serialize()` function

Reply:
We can't use the new `serialize()` function because it is nested object.
This function will be improved later see #3229